gusto/model.py

```diff
@@ -0,0 +1,228 @@
+from gusto.linear_solvers import ShallowWaterSolver, CompressibleSolver
+from gusto.state import State
+from gusto.transport_equation import VectorInvariant, AdvectionEquation, SUPGAdvection
+
+
+class Model(object, metaclass=ABCMeta):
+
+    eqn_opts = defaultdict(dict)
+
+    def __init__(self,
+                 mesh,
+                 timestepping=None,
+                 output=None,
+                 parameters=None,
+                 diagnostics=None,
+                 diagnostic_fields=None):
+
+        # store timestepping class
+        self.timestepping = timestepping
+
+        # update physical parameters
+        if parameters is not None:
+            for k, v in parameters.items():
+                self.parameters.__setattr__(k, v)
+
+        # create state
+        self.state = State(mesh=mesh,
+                           vertical_degree=self.vertical_degree,
+                           horizontal_degree=self.horizontal_degree,
+                           family=self.family,
+                           fieldlist=self.fieldlist,
+                           output=output,
+                           diagnostics=diagnostics,
+                           diagnostic_fields=diagnostic_fields)
+
+        # figure out if we're on a sphere
+        try:
+            self.on_sphere = (mesh._base_mesh.geometric_dimension() == 3 and mesh._base_mesh.topological_dimension() == 2)
+        except AttributeError:
+            self.on_sphere = (mesh.geometric_dimension() == 3 and mesh.topological_dimension() == 2)
+
+        #  build the vertical normal and define perp for 2d geometries
+        dim = mesh.topological_dimension()
+        if self.on_sphere:
+            x = SpatialCoordinate(mesh)
+            R = sqrt(inner(x, x))
+            self.state.k = interpolate(x/R, mesh.coordinates.function_space())
+            if dim == 2:
+                outward_normals = CellNormal(mesh)
+                self.state.perp = lambda u: cross(outward_normals, u)
+        else:
+            kvec = [0.0]*dim
+            kvec[dim-1] = 1.0
+            self.state.k = Constant(kvec)
+            if dim == 2:
+                self.state.perp = lambda u: as_vector([-u[1], u[0]])
+
+    @abstractproperty
+    def family(self):
+        pass
+
+    @abstractproperty
+    def vertical_degree(self):
+        pass
+
+    @abstractproperty
+    def horizontal_degree(self):
+        pass
+
+    @abstractproperty
+    def fieldlist(self):
+        pass
+
+    @abstractproperty
+    def parameters(self):
+        pass
+
+    @abstractproperty
+    def linear_solver(self):
+        pass
+
+    @abstractproperty
+    def forcing(self):
+        pass
+
+    @abstractproperty
+    def default_field_equations(self):
+        pass
+
+    @abstractproperty
+    def default_advection_schemes(self):
+        pass
+
+    def setup(self, linear_solver=None, forcing=None, field_equations=None,
+              advection_schemes=None):
```

Comment:
In general, these `setup` functions smell bad to me.  What appears to be going on is:

1. Create a `Model` instance (calls `__init__`)
1a. You can't use it for everything yet!
2a. Poke some things on the instance
2b. Poke some things on the member variables of the instance (e.g. `state`, to set up more stuff).
2c. You still can't use it!
3. Call `setup` on the model instance, in the process poking it some more. (e.g. providing advection schemes).
4. Now you can use it!


I think you could take advantage of `firedrake.utils.cached_property` to simplify a lot of stuff, and make it clearer what is going on.  By which I mean, `__init__` takes all things necessary to set up the instance, and then use `@cached_property` to provide "on-demand" computed properties (like the `linear_solver` and so forth).

That way you don't shove all of the complexity into this second `setup` phase, that I have to remember to call!

Reply:
yes, I see that makes things nicer because there is no need to call `model.setup`... let me try something like this...